### ISSUE TITLE
symlink: don’t call os.RemoveAll twice

### ISSUE THREAD

**Issue by stapelberg** [google/renameio#18]
This is not necessarily safe—a new tempdir might have taken place of the old one in between the regular os.RemoveAll call and the deferred one.

**Comment by twpayne** [comment]
Arguably the same could happen between the creation of the temp dir at line 159 and its removal in the defer.

**Comment by twpayne** [comment]
Presumably the only way to do this safely is to hang onto a file descriptor to the temporary directory and use that file descriptor to remove the directory? However, to my knowledge, there is no system call to remove a directory given only an open fd to it. I hope I'm wrong.

**Comment by stapelberg** [comment]
Not sure I follow? Are you saying we shouldn’t fix this issue, or are you saying fixing the issue won’t be enough? Why?

**Comment by stapelberg** [comment]
Wait, are you talking about $TMPDIR changing? That’s not what I’m talking about.

**Comment by twpayne** [comment]
No, I'm talking about the a temp dir being created in the `Symlink` function, that temp dir being replaced by another one with the same name, and then the second call to `os.RemoveAll` removing the new temp dir rather than the original one.

**Comment by stapelberg** [comment]
…but a new temp dir can only take the old one’s place once the old one was removed? The scenario you’re describing doesn’t seem plausible to me.

**Comment by twpayne** [comment]
OK, so I've misunderstood the issue :)

**Comment by twpayne** [comment]
My understanding was that these two lines tried to remove the same directory:
https://github.com/google/renameio/blob/master/tempfile.go#L163
https://github.com/google/renameio/blob/master/tempfile.go#L174

**Comment by stapelberg** [comment]
Okay :) The series of steps that I have in mind are:

* pid 23: ioutil.TempDir creates e.g. /tmp/xyz.123
* pid 23: regular os.RemoveAll removes /tmp/xyz.123
* pid 42: ioutil.TempDir creates /tmp/xyz.123 by extreme chance
* pid 23: deferred os.RemoveAll removes /tmp/xyz.123 again
* pid 42 malfunctions

**Comment by twpayne** [comment]
OK, I think we're on the same page here.

So the question is: how can pid 23 be sure that it removes the exact same directory that it created?

**Comment by stapelberg** [comment]
By calling os.RemoveAll only once, i.e. skipping the deferred call if err==nil when returning.

**Comment by twpayne** [comment]
[This comment](https://github.com/google/renameio/issues/18#issuecomment-586761133) referred to the possible sequence:

* pid 23: ioutil.TempDir creates e.g. /tmp/xyz.123
* pid 59: removes /tmp/xyz.123
* pid 42: creates /tmp/xyz.123 by extreme chance
* pid 23: regular or deferred os.RemoveAll removes /tmp/xyz.123
* pid 23 malfunctions
* pid 42 malfunctions

**Comment by twpayne** [comment]
> By calling os.RemoveAll only once, i.e. skipping the deferred call if err==nil when returning.

I agree that this is an improvement on the existing code :)

**Comment by stapelberg** [comment]
Yeah, that’s more in the malicious category, not in the “two processes just kept running for years and suddenly raced with each other” category :)

**Comment by twpayne** [comment]
If you could keep a handle on the actual directory (rather than just its name) then you could remove it safely. But this would require a bit more work than the clear improvement that you're already proposing here. I'll keep quiet now :)

**Comment by twpayne** [comment]
FWIW, my quick search through Linux system calls did not reveal any system call that allows you to remove either a file or directory given its file descriptor alone. As far as I can tell, all other relevant syscalls operate on paths alone.

In practice, this probably means that the only hope we have is generating very unique never-to-be-repeated paths for temporary directories.

> the “two processes just kept running for years and suddenly raced with each other” category

I'd love to hear the story behind this. Sounds juicy :)

**Comment by stapelberg** [comment]
> FWIW, my quick search through Linux system calls did not reveal any system call that allows you to remove either a file or directory given its file descriptor alone. As far as I can tell, all other relevant syscalls operate on paths alone.

Does https://manpages.debian.org/buster/manpages-dev/unlinkat.2.en.html not fit the bill?

That said, I think that’s all going a bit overboard :)

> I'd love to hear the story behind this. Sounds juicy :)

No story here (that I know of). The (mostly theoretical) issue was reported via mail.

**Comment by twpayne** [comment]
Nice find! I failed to find `unlinkat` in my searches. It looks perfect.

Maybe it's going a bit overboard, but isn't it nice to know that you've eliminated a whole bunch of will-probably-never-ever-happen-but-would-drive-you-insane-trying-to-find-the-root-cause-if-they-ever-did bugs?

**Comment by stapelberg** [comment]
> Maybe it's going a bit overboard, but isn't it nice to know that you've eliminated a whole bunch of will-probably-never-ever-happen-but-would-drive-you-insane-trying-to-find-the-root-cause-if-they-ever-did bugs?

In theory yes, but not with an unproportionate amount of code (compared to the rest of renameio). I have dealt with a custom os.RemoveAll recently, and it’s not trivial.
